File: clients/minecraft/jvm/paper/src/main/java/de/rafael/atomic/cloud/heart/Heart.java
package de.rafael.atomic.cloud.heart;

import de.rafael.atomic.cloud.CloudPlugin;
import io.papermc.paper.threadedregions.scheduler.ScheduledTask;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import lombok.RequiredArgsConstructor;
import org.bukkit.Bukkit;

@RequiredArgsConstructor
public class Heart implements Consumer<ScheduledTask> {

  private final long interval;
  private final AtomicBoolean running = new AtomicBoolean(false);

  public void start() {
    CloudPlugin.LOGGER.info("Starting heart of this server...");
    running.set(true);
    Bukkit.getAsyncScheduler()
        .runAtFixedRate(CloudPlugin.INSTANCE, this, 0, interval, TimeUnit.SECONDS);
  }

  public void stop() {
    CloudPlugin.LOGGER.info("Stopping heart of this server...");
    running.set(false);
  }

  @Override
  public void accept(ScheduledTask scheduledTask) {
    if (!running.get()) {
      scheduledTask.cancel();
      return;
    }
    CloudPlugin.INSTANCE.getConnection().beatHeart();
  }
}

File: clients/minecraft/jvm/paper/src/main/kotlin/de/rafael/atomic/cloud/heart/Heart.kt
package de.rafael.atomic.cloud.heart

import de.rafael.atomic.cloud.Plugin
import io.papermc.paper.threadedregions.scheduler.ScheduledTask
import org.bukkit.Bukkit
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

object Heart {

    private const val HEARTBEAT_INTERVAL_SEC = 10L // Every 10 seconds
    private val shouldBeat = AtomicBoolean(true)

    fun start() {
        Plugin.logger.info("Starting heart of this server...")
        shouldBeat.set(true)
        Bukkit.getAsyncScheduler().runAtFixedRate(Plugin, {
            beat(it)
        }, 0, HEARTBEAT_INTERVAL_SEC, TimeUnit.SECONDS)
    }

    fun stop() {
        Plugin.logger.info("Stopping heart of this server...")
        shouldBeat.set(false)
    }

    private fun beat(task: ScheduledTask) {
        if (!shouldBeat.get()) {
            task.cancel()
            return
        }
        TODO("Not yet implemented")
    }
}
